zerver/lib/message.py

```diff
@@ -1317,12 +1326,21 @@ def apply_unread_message_event(
             topic=topic,
         )
 
-        if (
-            stream_id not in state["muted_stream_ids"]
-            # This next check hits the database.
-            and not topic_has_visibility_policy(
-                user_profile, stream_id, topic, UserTopic.VisibilityPolicy.MUTED
+        stream_muted = stream_id in state["muted_stream_ids"]
+        try:
+            user_topic = UserTopic.objects.get(
+                user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic
             )
+            visibility_policy = user_topic.visibility_policy
+        except UserTopic.DoesNotExist:
+            visibility_policy = UserTopic.VisibilityPolicy.INHERIT
+        # A stream message is unmuted if it belongs to:
+        # * a not muted topic in a normal stream
+        # * an unmuted or followed topic in a muted stream
+        if (not stream_muted and visibility_policy != UserTopic.VisibilityPolicy.MUTED) or (
```

Comment:
This first part might be better as just `visibility_policy != UserTopic.VisibilityPolicy.MUTED` -- it's not actually important whether the stream is muted if the topic policy is `MUTED`, and that would I think make this a bit easier to read.

Reply:
When `visibility_policy = INHERIT`, using only `visibility_policy != UserTopic.VisibilityPolicy.MUTED` doesn't cover the case :

A message in a normal topic in a muted stream is not treated as unmuted message.

We can re-write it as
```
if not ((visibility_policy == MUTED) or (stream_muted and visibility_policy == INHERIT)):
   state["unmuted_stream_msgs"].add(message_id)
```

But I think, the existing logic is more readable ? We can change that if the later seems better to you.

Reply:
Good catch, I like the logic you have, thanks for clarifying the bug my proposal would introduce :).